File: wrappers/julia/src/encoding.jl
"""
    Encoding

A loaded BPE encoding ready for tokenization operations.
"""
struct Encoding
    name::String
    spec::EncodingSpec
    rank_payload::Vector{UInt8}
end

"""
    encode(enc::Encoding, text::AbstractString) -> Vector{UInt32}

Encode text into BPE token IDs.
"""
encode(enc::Encoding, text::AbstractString)::Vector{UInt32} =
    ffi_encode_bpe(enc.rank_payload, text)

"""
    decode(enc::Encoding, tokens::Vector{UInt32}) -> String

Decode BPE token IDs back into a UTF-8 string.
"""
decode(enc::Encoding, tokens::Vector{UInt32})::String =
    ffi_decode_bpe(enc.rank_payload, tokens)

"""
    count(enc::Encoding, text::AbstractString) -> Int

Count the number of BPE tokens in text without materializing the token array.
"""
count(enc::Encoding, text::AbstractString)::Int =
    ffi_count_bpe(enc.rank_payload, text)

"""
    count_tokens(enc::Encoding, text::AbstractString) -> Int

Alias for `count`.
"""
const count_tokens = count

"""
    is_within_token_limit(enc::Encoding, text::AbstractString, limit::Int) -> Union{Int, Nothing}

Check if text is within a token limit.
Returns the token count if within limit, `nothing` if exceeded.
"""
is_within_token_limit(enc::Encoding, text::AbstractString, limit::Int)::Union{Int, Nothing} =
    ffi_is_within_token_limit(enc.rank_payload, text, limit)

"""
    encode_chat(enc::Encoding, messages::Vector{ChatMessage}; kwargs...) -> Vector{UInt32}

Encode chat messages into BPE token IDs.
"""
function encode_chat(enc::Encoding, messages::Vector{ChatMessage}; kwargs...)::Vector{UInt32}
    opts = ChatOptions(; kwargs...)
    text = format_chat_messages(messages; opts=opts)
    return encode(enc, text)
end

"""
    count_chat(enc::Encoding, messages::Vector{ChatMessage}; kwargs...) -> Int

Count BPE tokens for chat messages.
"""
function count_chat(enc::Encoding, messages::Vector{ChatMessage}; kwargs...)::Int
    opts = ChatOptions(; kwargs...)
    text = format_chat_messages(messages; opts=opts)
    return count(enc, text)
end

"""
    is_chat_within_token_limit(enc::Encoding, messages::Vector{ChatMessage}, limit::Int; kwargs...) -> Union{Int, Nothing}

Check if chat messages are within a token limit.
"""
function is_chat_within_token_limit(enc::Encoding, messages::Vector{ChatMessage}, limit::Int; kwargs...)::Union{Int, Nothing}
    opts = ChatOptions(; kwargs...)
    text = format_chat_messages(messages; opts=opts)
    return is_within_token_limit(enc, text, limit)
end

"""
    encode_file_path(enc::Encoding, path::AbstractString) -> Vector{UInt32}

Encode a file's contents into BPE token IDs.
"""
encode_file_path(enc::Encoding, path::AbstractString)::Vector{UInt32} =
    ffi_encode_bpe_file(enc.rank_payload, path)

"""
    count_file_path(enc::Encoding, path::AbstractString) -> Int

Count BPE tokens in a file.
"""
count_file_path(enc::Encoding, path::AbstractString)::Int =
    ffi_count_bpe_file(enc.rank_payload, path)

"""
    is_file_path_within_token_limit(enc::Encoding, path::AbstractString, limit::Int) -> Union{Int, Nothing}

Check if a file's content is within a token limit.
"""
is_file_path_within_token_limit(enc::Encoding, path::AbstractString, limit::Int)::Union{Int, Nothing} =
    ffi_is_within_token_limit_file(enc.rank_payload, path, limit)

function Base.show(io::IO, enc::Encoding)
    print(io, "Encoding(\"$(enc.name)\", n_vocab=$(enc.spec.n_vocab))")
end

File: wrappers/php/src/Encoding.php
<?php

declare(strict_types=1);

namespace TurboToken;

class Encoding
{
    private string $rankPayload;
    private EncodingSpec $spec;

    public function __construct(string $rankPayload, EncodingSpec $spec)
    {
        $this->rankPayload = $rankPayload;
        $this->spec = $spec;
    }

    public function name(): string
    {
        return $this->spec->name;
    }

    public function nVocab(): int
    {
        return $this->spec->nVocab;
    }

    public function eotToken(): int
    {
        return $this->spec->eotToken();
    }

    /**
     * @return int[]
     */
    public function encode(string $text): array
    {
        return NativeBridge::encodeBpeFromRanks($this->rankPayload, $text);
    }

    /**
     * @param int[] $tokens
     */
    public function decode(array $tokens): string
    {
        return NativeBridge::decodeBpeFromRanks($this->rankPayload, $tokens);
    }

    public function count(string $text): int
    {
        return NativeBridge::countBpeFromRanks($this->rankPayload, $text);
    }

    public function countTokens(string $text): int
    {
        return $this->count($text);
    }

    /**
     * Returns token count if within limit, null if exceeded.
     */
    public function isWithinTokenLimit(string $text, int $limit): ?int
    {
        return NativeBridge::isWithinTokenLimitBpeFromRanks($this->rankPayload, $text, $limit);
    }

    /**
     * @param ChatMessage[] $messages
     * @param array{template?: string} $opts
     * @return int[]
     */
    public function encodeChat(array $messages, array $opts = []): array
    {
        $template = ChatTemplate::resolveTurboTokenV1();
        $text = $template->formatMessages($messages);
        return $this->encode($text);
    }

    /**
     * @param ChatMessage[] $messages
     * @param array{template?: string} $opts
     */
    public function countChat(array $messages, array $opts = []): int
    {
        $template = ChatTemplate::resolveTurboTokenV1();
        $text = $template->formatMessages($messages);
        return $this->count($text);
    }

    /**
     * @param ChatMessage[] $messages
     * @param array{template?: string} $opts
     */
    public function isChatWithinTokenLimit(array $messages, int $limit, array $opts = []): ?int
    {
        $template = ChatTemplate::resolveTurboTokenV1();
        $text = $template->formatMessages($messages);
        return $this->isWithinTokenLimit($text, $limit);
    }

    /**
     * @return int[]
     */
    public function encodeFilePath(string $filePath): array
    {
        return NativeBridge::encodeBpeFileFromRanks($this->rankPayload, $filePath);
    }

    public function countFilePath(string $filePath): int
    {
        return NativeBridge::countBpeFileFromRanks($this->rankPayload, $filePath);
    }

    public function isFilePathWithinTokenLimit(string $filePath, int $limit): ?int
    {
        return NativeBridge::isWithinTokenLimitBpeFileFromRanks($this->rankPayload, $filePath, $limit);
    }
}
